### PR TITLE
Fix use-primary-constructor in the presence of side-effects on the parameters.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/70868

### CODE REVIEW

File: src/Analyzers/CSharp/Tests/UsePrimaryConstructor/UsePrimaryConstructorTests.cs [hunk 1]
```diff
@@ -3291,4 +3291,296 @@ partial class C<T> : Base
...
+                        A = ++value1;
+                        B = ++value2;
+                    }
+                }
+                """,
+            FixedCode = """
+                class C(int value1, int value2)
+                {
+                    public int B = ++value2;
+                    public int A = ++value1;
```
Comment (RikkiGibson) [inline src/Analyzers/CSharp/Tests/UsePrimaryConstructor/UsePrimaryConstructorTests.cs:3522]: What behavior would be expected in a case like the following?
```cs
class Pair
{
    public int First;
    public int Second;

    public Pair(Stack<int> items)
    {
        Second = items.Pop();
        First = items.Pop();
    }
}
```

Comment (CyrusNajmabadi) [reply src/Analyzers/CSharp/Tests/UsePrimaryConstructor/UsePrimaryConstructorTests.cs:3522]: @RikkiGibson badness :)

**Review by akhera99 — APPROVED** [review]

File: src/Analyzers/CSharp/Tests/UsePrimaryConstructor/UsePrimaryConstructorTests.cs (hunk 1 above)
Comment (RikkiGibson) [reply src/Analyzers/CSharp/Tests/UsePrimaryConstructor/UsePrimaryConstructorTests.cs:3522]: I didn't mean to imply the fixer behavior should be adjusted from what you already have based on the possibility of this scenario. More that I think it's good to have a couple tests here and there for scenarios which show where the "boundaries" of the fixer's heuristics are. However, I also don't see a need to go back and add the test now.
